apm-agent-plugins/apm-vertx/apm-vertx4-plugin/src/main/java/co/elastic/apm/agent/vertx/v4/web/WebInstrumentation.java

```diff
@@ -0,0 +1,278 @@
+import io.vertx.core.spi.tracing.VertxTracer;
+import io.vertx.ext.web.RoutingContext;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.NamedElement;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.nameContains;
+import static net.bytebuddy.matcher.ElementMatchers.nameStartsWith;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.namedOneOf;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesNoArguments;
+
+@SuppressWarnings("JavadocReference")
+public abstract class WebInstrumentation extends Vertx4Instrumentation {
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("vertx", "vertx-web", "experimental");
+    }
+
+    /**
+     * Instruments {@link io.vertx.core.impl.ContextImpl#tracer}} to return a noop tracer in case no tracer has been specified.
+     */
+    public static class ContextImplTracerInstrumentation extends WebInstrumentation {
+        @Override
+        public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return named("io.vertx.core.impl.ContextImpl")
+                .and(not(isInterface()));
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("tracer").and(takesNoArguments());
+        }
+
+        @Override
+        public String getAdviceClassName() {
+            return "co.elastic.apm.agent.vertx.v4.web.WebInstrumentation$ContextImplTracerInstrumentation$ContextImplTracerAdvice";
+        }
+
+        public static class ContextImplTracerAdvice {
+
+            @AssignTo.Return
+            @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
+            public static VertxTracer<?, ?> receiveRequest(@Advice.Return @Nullable VertxTracer<?, ?> currentTracer) {
+                return currentTracer == null ? NoopVertxTracer.INSTANCE : currentTracer;
+            }
+        }
+
+    }
+
+    public static abstract class VertxTracerInstrumentation extends WebInstrumentation {
+
+        @Override
+        public final ElementMatcher<? super NamedElement> getTypeMatcherPreFilter() {
+            return nameContains("Tracer");
+        }
+
+        @Override
+        public final ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return hasSuperType(named("io.vertx.core.spi.tracing.VertxTracer"))
```

Comment:
[question] Does it means we instrument any existing tracer if there is any and if there is none we inject our own no-op which comes not-so-no-op anymore once it's instrumented ?

Reply:
exactly

Reply:
In that case if you have time to add a minor comment to explain that a bit and make it more obvious that would really be nice. I'm not sure that I'll remember that trick in a few months :-).